keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,723 @@
+
+If a lien is set to Deployment A, only a deletion request to Deployment A is blocked.
+Users can still request to delete Replicaset B and Pod C directly, but they can't request to delete them through Deployment A.
+
+If a lien is set to Replicaset B, only a deletion request to Replicaset B is blocked.
+Users can still request to delete Deployment A and Pod C directly.
+When deletion of the Deployment A is requested, whether it completes immediately or not depends on the cascading policy and Replicaset B's `blockOwnerDeletion`.
+The behaviors are as follows:
+
+- Foreground cascading deletion:
+  - `blockOwnerDeletion=true`: Deletion of Deployment A isn't completed until Replicaset B is deleted
+  - otherwise: Deployment A is deleted immediately
+- Background cascading deletion: Deployment A is deleted immediately
+
+#### Namespace Deletion
+
+It may also not be clear how [namespace deletion](https://kubernetes.io/docs/tasks/administer-cluster/namespaces/#deleting-a-namespace) behaves with lien.
+Therefore, this section describes the behavior.
+
+In summary, lien will block the namespace deletion, but would not block deletion of any resources in the namespace individually.
+
+If a lien is set to a namespace, only a deletion request to the namespace is blocked.
+Users can still request to delete each object in the namespace directly, but they can't request to delete it through the namespace deletion.
+
+If a lien is set to one of objects in a namespace, only a deletion request to the object is blocked.
+Users can still request to delete its namespace or other objects in the namespace.
+When deletion of the namespace A is requested, the deletion isn't completed until the object with lien is deleted.
+
+#### Block Adding Additional `Liens` while Deleting
+
+`Liens` shouldn't be added after `DeletionTimestamp` is non-nil, which means pre-deletion processes for finalizers are being handled.
+Adding `Liens` while deleting itself won't do any harm for the cluster, because the resource will be deleted without any additional deletion requests that will be blocked by `Liens`.
+However, users may think that the successful addition of the `Liens` means that the resource won't be deleted until the `Liens` are deleted, which isn't true.
+To avoid such a misunderstanding from users, a request for adding additional `Liens` to a resource with non-nil `DeletionTimestamp` should be blocked by api-server.
+
+#### Per Controller `Liens` vs Per Reason `Liens`
+
+The same controllers may need to add `Liens` to a certain resource for multiple reasons, like multiple resources are using it.
+In such a case, if `Liens` are added per controller basis, such a controller needs to be sure that no other resources are actually using the resource, when removing its `Liens` from the resource.
+On the other hand, if `Liens` are added per reason basis, such a controller can remove the `Lien` for the resource, without considering other resources.
```

Comment:
So (again, tell me if I am wrong) you're saying it's not like finalizer which is just saying "wait", but it really does model a dependency graph?

Reply:
maybe pseudocode for such a controller would help?

Reply:
This *is* effectively a dep graph.  See above comments about the size of this list.  a `[]string` API here means that we can put an effective limit like "only 256 liens" allowed and add monitoring and see if that is actually a problem limit.  once you enumerate the lien-holders, the limit MUST be larger, which puts the whole system at risk.

Reply:
> I'm personally against having effectively a lien for every single "referencee".

I agree with you that adding a big list isn't good from a scalability view point.

On the other hand, adding a list of used resources will give users a better understanding on which resources are actually using it. In other words, it will let users know which other resources should be deleted to delete the resource safely. It will be a big benefit from a UX view point.

In addition, from the lien consumers' implementation view point, checking all the potential using resources before deleting a lien will be a time consuming operation. For example, in secret protection case, the controller needs to loop for all the secret to check whether any Pods or PVCs are using the secret, in order to delete the per controller lien (Per object lien will allow controllers just to check if the resource that is referenced from per object lien is still using the Secret).

Actaully, pv-protection and pvc-protection are using the similar approach to the per controller lien case above by using finalizer. However, it just checks whether any Pods or PVCs are using the secret after controller observes non-nil DeletionTimeStamp on the secret, or only after users request to delete it. Similar approach won't be used for lien, because the lien blocks deletion request itself and there will be no room to add such a check for controllers. Therefore, controllers need to keep the lien up-to-date all the time. (For secret-protection to take the similar approach to finalizer, it needs to be implemented as a validation webhook for deletion that checks whether any Pods or PVCs are using the secret, instead of using lien. It means that lien becomes useless for this kind of use cases).

Reply:
> I will consider whether we can set an effective limit or there will be other good approaches.

I'm considering if there is a way that consumers can handle per controller lien efficiently.

@wojtek-t 's comment [here](https://github.com/kubernetes/enhancements/pull/2640#discussion_r703553386) let me think whether current secret-protection-controller logic with per object lien is really simple (Whether creating an in-memory dep-graph in such controllers, like the garbage collector does, might be better, and so on. Per object lien may be still better from the UX view point above, though).

Any feedback and ideas are welcome, as we are running out of time.

Reply:
> In addition, from the lien consumers' implementation view point, checking all the potential using resources before deleting a lien will be a time consuming operation. 

I kind of disagree.
A controller may keep the in-memory mapping from secret to list of objects that are referencing it.

Let's for a moment assume that only Pods can reference it - this trivially generalizes for other objects.
1. Updating that mapping is simple - whenever some Pod is created/updated/delete - we simply update the mapping for all secrets it is referencing.
[FWIW - we use a similar approach in secret-manager in kubelet.]
2. Whenever we change the number of references from 0 to more than 0 or vice versa, we update the lien of that secret.

So that's actually fairly efficient implementation that we can have.

BTW - another advantage of that is that we significantly reduce the number of API calls (as we don't have to update the secret on every new reference to it) - which is another huge advantage of this approach.

Reply:
It seems that there will be a way to efficiently update lien by consumer side, so I will update the KEP to only allow per controller basis. Thank you for your review and sharing the ideas in detail. @wojtek-t